tensorflow/java/src/main/java/org/tensorflow/Server.java

```diff
@@ -0,0 +1,82 @@
+/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+package org.tensorflow;
+
+/**
+ * An in-process TensorFlow server, for use in distributed training.
+ *
+ * A {@code tf.train.Server} instance encapsulates a set of devices and a
+ * {@code tf.Session} target that can participate in distributed training. A
+ * server belongs to a cluster (specified by a {@code tf.train.ClusterSpec}),
+ * and corresponds to a particular task in a named job. The server can
+ * communicate with any other server in the same cluster.
+ *
+ * <p><b>WARNING:</b>A {@code Server} owns resources that <b>must</b> be
+ * explicitly freed by invoking {@link #close()}.
+ *
+ * <p>Instances of a {@code Server} are thread-safe.
+ */
+public final class Server implements AutoCloseable {
+
+  /** 
+   * Constructs a new instance of server. 
+   *
+   * @param config Server definition specified as a serialized
+   *        <a href="https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/protobuf/tensorflow_server.proto">ServerDef</a>
+   *        protocol buffer.
+   */
+  public Server(byte[] serverDef) {
+    nativeHandle = allocate(serverDef);
+  }
+
+  /** Starts this server. */
+  public synchronized void start() {
+    start(nativeHandle);
+  }
+
+  /** Stops this server. */
+  public synchronized void stop() {
+    stop(nativeHandle);
+  }
+
+  /** Blocks until the server has shut down (currently blocks forever). */
+  public synchronized void join() {
```

Comment:
Given that `join()` blocks, are you sure we want to mark this method as `synchronized`?
It appears to me, that if one thread were to invoke `join()` first, then no other thread will be able to invoke `stop()` and as a result both threads will remain blocked forever. Am I missing something?

Reply:
I've checked code and looks like `Server` methods `start`, `stop` and `join` are synchronized on underlying layer (I didn't know that). So, we only need to prohibit parallel calls of `delete` and other functions. Do do that I added rw-lock, so please have a look.

Using this lock, user will be able to call `stop` method from other thread when current thread is locked by `join` method.

Reply:
By the way, I checked calling `stop` when server is locked by `join` and looks like it doesn't work (see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc#L412)

Reply:
I investigated a bit more and found out that `close()` just can't work correctly. `Server` destructor calls `stop()` wrapped into `TF_CHECK_OK`, but `stop()` returns `unimplemented` exception, as result `TF_CHECK_OK` crushes the process using sig 6.

Looks like it should be fixed, but in other PR (it's not about Java API at all).